### PR TITLE
Output validator [OutputValidator v1.0.0]

### PR DESCRIPTION
**Pull request by 0xDEnYO**
# Which Jira task belongs to this PR?
https://lifi.atlassian.net/browse/LF-15125

# Why did I implement it this way?

# Checklist before requesting a review

- [x] I have performed a self-review of my code
- [x] This pull request is as small as possible and only tackles one problem
- [ ] I have added tests that cover the functionality / test the bug
- [ ] For new facets: I have checked all points from this list: https://www.notion.so/lifi/New-Facet-Contract-Checklist-157f0ff14ac78095a2b8f999d655622e
- [ ] I have updated any required documentation

# Checklist for reviewer (DO NOT DEPLOY and contracts BEFORE CHECKING THIS!!!)

- [ ] I have checked that any arbitrary calls to external contracts are validated and or restricted
- [ ] I have checked that any privileged calls (i.e. storage modifications) are validated and or restricted
- [ ] I have ensured that any new contracts have had AT A MINIMUM 1 preliminary audit conducted on <date> by <company/auditor>

### CODE REVIEW

File: script/helperFunctions.sh [hunk 1]
```diff
@@ -3616,6 +3613,54 @@ function convertToBcInt() {
   echo "$1" | tr -d '\n' | bc
 }
 
+# function extractDeployedAddressFromRawReturnData() {
```
Comment (0xDEnYO) [inline script/helperFunctions.sh]: I wanna keep this for now as I need to test this thoroughly

File: docs/OutputValidator.md [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+2. The contract always returns the expected amount to the calling contract using `LibAsset.transferAsset`
+3. **Assumes positive slippage**: Always transfers excess to validation wallet (handles zero excess by transferring 0 tokens)
+
+### ERC20 Token Flow
+
+1. The calling contract (Diamond) must have sufficient ERC20 token balance
+2. The OutputValidator checks the Diamond's ERC20 balance using `ERC20(tokenAddress).balanceOf(msg.sender)`
+3. **Assumes positive slippage**: Always transfers excess to validation wallet (handles zero excess by transferring 0 tokens)
+4. The Diamond retains the expected amount
+
```
[the remark this answers is not in the document]
Comment (0xDEnYO) [reply docs/OutputValidator.md]: this is by design and intended as-is

File: docs/OutputValidator.md [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+### Native Token Flow
+
+1. The calling contract (Diamond) sends a portion of native tokens as `msg.value` for excess handling
+2. **Calculates excess**: `excessAmount = (contract_balance + msg.value) - expectedAmount`
+3. **Smart distribution**:
+   - If `excessAmount >= msg.value`: All `msg.value` goes to validation wallet (contract balance covers expected amount)
+   - If `excessAmount < msg.value`: Sends `excessAmount` to validation wallet, returns `msg.value - excessAmount` to sender
+4. **User receives expected amount** through the normal swap flow, not from this contract
+
```
[the remark this answers is not in the document]
Comment (0xDEnYO) [reply docs/OutputValidator.md]: adjusted

File: src/Periphery/OutputValidator.sol [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        // make sure we do not attempt any token transfers if there is no excess amount
+        if (excessAmount > 0) {
+            // validate wallet address
+            if (validationWalletAddress == address(0)) {
+                revert InvalidCallData();
+            }
+
+            // transfer excess to validation wallet
+            // no need to validate the tokenAddress, it will fail if it's an invalid address
+            tokenAddress.safeTransferFrom(
```
Comment (0xDEnYO) [inline src/Periphery/OutputValidator.sol]: we should be using LibAsset here but the ERC20 transfer function is private so we cant call it directly. This is being changed by another PR (LDA 2.0) already. And for this contract here this is the most gas-efficient way to do the transfer

File: docs/OutputValidator.md [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+- Zero amounts and balances
+- Various slippage scenarios
+
+### **Test Statistics**
+
+- **Comprehensive test coverage** covering all code paths
+- **All branches covered** including edge cases
+- **Realistic scenarios** using MockDEX and Diamond integration
+- **100% unit test coverage** as per project requirements
+
```
[the remark this answers is not in the document]
Comment (0xDEnYO) [reply docs/OutputValidator.md]: removed

File: src/Periphery/OutputValidator.sol [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+                expectedAmount
+            );
+
+            // transfer excess to validation wallet
+            LibAsset.transferAsset(
+                tokenAddress,
+                payable(validationWalletAddress),
+                actualAmount - expectedAmount
+            );
+        } else {
```
[the remark this answers is not in the document]
Comment (0xDEnYO) [reply src/Periphery/OutputValidator.sol]: outdated

File: deployments/arbitrum.diamond.staging.json [hunk 7]
```diff
@@ -191,7 +191,8 @@
       "ReceiverStargateV2": "",
       "RelayerCelerIM": "0xa1Ed8783AC96385482092b82eb952153998e9b70",
       "Patcher": "0x3971A968c03cd9640239C937F8d30D024840E691",
-      "TokenWrapper": "0xF63b27AE2Dc887b88f82E2Cc597d07fBB2E78E70"
+      "TokenWrapper": "0xF63b27AE2Dc887b88f82E2Cc597d07fBB2E78E70",
+      "OutputValidator": "0x266f7a44969d0003AFD94cdf96bdfdEDd8754133"
     }
```
[the remark this answers is not in the document]
Comment (0xDEnYO) [reply deployments/arbitrum.diamond.staging.json:238]: not true, it's there

File: deployments/optimism.staging.json [hunk 8]
```diff
@@ -45,5 +45,6 @@
   "LidoWrapper": "0x462A9B6879770050021823D63aE62470E65Af8D4",
   "Permit2Proxy": "0x808eb38763f3F51F9C47bc93Ef8d5aB7E6241F46",
   "GasZipFacet": "0xfEeCe7B3e68B9cBeADB60598973704a776ac3ca1",
+  "OutputValidator": "0x266f7a44969d0003AFD94cdf96bdfdEDd8754133",
   "GlacisFacet": "0x36e1375B0755162d720276dFF6893DF02bd49225"
```
[the remark this answers is not in the document]
Comment (0xDEnYO) [reply deployments/optimism.staging.json]: image
image


it's all there mate. Why do you not see these?

File: deployments/arbitrum.diamond.staging.json [hunk 9]
```diff
@@ -195,7 +195,10 @@
       "ReceiverAcrossV4": "0x8fd9e1893fdE2e52b49F2875E63277Ba5675D014",
       "ReceiverChainflip": "",
       "ReceiverStargateV2": "",
-      "TokenWrapper": "0xF63b27AE2Dc887b88f82E2Cc597d07fBB2E78E70"
+      "RelayerCelerIM": "0xa1Ed8783AC96385482092b82eb952153998e9b70",
+      "Patcher": "0x3971A968c03cd9640239C937F8d30D024840E691",
+      "TokenWrapper": "0xF63b27AE2Dc887b88f82E2Cc597d07fBB2E78E70",
+      "OutputValidator": "0x266f7a44969d0003AFD94cdf96bdfdEDd8754133"
```
[the remark this answers is not in the document]
Comment (0xDEnYO) [reply deployments/arbitrum.diamond.staging.json]: there's only one Patcher entry in that file

File: deployments/optimism.diamond.staging.json [hunk 10]
```diff
@@ -179,7 +179,10 @@
       "ReceiverAcrossV4": "0x1d5bD612Ce761060A4bEd77b606ab7e723D4E91E",
       "ReceiverChainflip": "",
       "ReceiverStargateV2": "",
-      "TokenWrapper": "0xF63b27AE2Dc887b88f82E2Cc597d07fBB2E78E70"
+      "RelayerCelerIM": "0xa1Ed8783AC96385482092b82eb952153998e9b70",
+      "TokenWrapper": "0xF63b27AE2Dc887b88f82E2Cc597d07fBB2E78E70",
+      "OutputValidator": "0x266f7a44969d0003AFD94cdf96bdfdEDd8754133",
+      "Patcher": ""
```
[the remark this answers is not in the document]
Comment (0xDEnYO) [reply deployments/optimism.diamond.staging.json]: removed

File: deployments/arbitrum.diamond.staging.json (hunk 9 above)
[the remark this answers is not in the document]
Comment (0xDEnYO) [reply deployments/arbitrum.diamond.staging.json]: yes I looked in the wrong file, correct. Fixed now.

**Review by mirooon — APPROVED** [review]
